### ISSUE TITLE
bash exclamation mark in pipe results in error but is undetected by shellcheck

### ISSUE THREAD

**Issue by programminghoch10** [koalaman/shellcheck#2810]
#### For bugs
- Rule Id: *(SC2143 in wider sense)*
- My shellcheck version (`shellcheck --version` or "online"): 0.9.0 locally and online version at time of writing
- [x] The rule's wiki page does not already cover this
- [x] I tried on https://www.shellcheck.net/ and verified that this is still a problem on the latest commit

#### For new checks and feature suggestions
- [x] https://www.shellcheck.net/ (i.e. the latest commit) currently gives no useful warnings about this
- [x] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:

ShellCheck warned me about this:
```bash
#!/bin/bash
[ -z "$(echo a | grep a)" ] && echo true
```
```
In test.sh line 1:
[ -z "$(echo a | grep a)" ] && echo true
  ^-- SC2143 (style): Use ! grep -q instead of comparing output with [ -z .. ].
```

so I followed the advice:
```bash
#!/bin/bash
echo a | ! grep a && echo true
```
which doesn't work:
```
Line 2: Syntax error at unexpected symbol »!«
Line 2: `echo a | ! grep a && echo true'
```
however ShellCheck doesn't recognize this issue at all:
```
No issues detected!
```

correct would be (i think, maybe depends on `pipefail` setting too):
```bash
#!/bin/bash
! echo a | grep a && echo true
```

**Comment by koalaman** [comment]
ShellCheck unintentionally interpreted the `!` as a string command name both in the second ! in `! ! cmd` and in `foo | ! bar`. This has now been fixed, and warnings have been added for each shell that doesn't support one or both. Thanks!
